### PR TITLE
Adding prefix to gateway internal metrics

### PR DESCRIPTION
**Pull request by krunaljain**

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -174,6 +176,10 @@ type GatewayConfig struct {
 	EtcdSnapCount    *uint64 `json:",omitempty"`
 	EtcdMaxSnapFiles *uint   `json:",omitempty"`
 	EtcdMaxWalFiles  *uint   `json:",omitempty"`
+
+	// Default reporting delay for gateway internal metrics
+	InternalMetricsReportingDelay         *string
```
Comment (mdubbyap) [inline config/config.go]: please mark explicitly as a json element like the above, also comment on what this does and how it is special

File: config/config.go [hunk 2]
```diff
@@ -174,6 +176,10 @@ type GatewayConfig struct {
...
+	InternalMetricsReportingDelayDuration *time.Duration
```
Comment (mdubbyap) [inline config/config.go]: should not be settable via json

File: main.go [hunk 3]
```diff
@@ -255,6 +257,16 @@ func setupListeners(tk timekeeper.TimeKeeper, hostname string, loadedConfig *con
 			return nil, errDupeListener
 		}
 		nameMap[name] = true
+		limitedLogger := &log.RateLimitedLogger{
```
Comment (mdubbyap) [inline main.go]: don't need to duplicated the logger

File: main.go [hunk 4]
```diff
@@ -582,9 +604,23 @@ func (p *gateway) scheduleStatCollection(ctx context.Context, scheduler *sfxclie
 	return finishedContext, cancelFunc
 }
 
-func (p *gateway) setupForwardersAndListeners(ctx context.Context, loader *config.Loader, loadedConfig *config.GatewayConfig, logger log.Logger, scheduler *sfxclient.Scheduler) (signalfx.Sink, map[string]http.Handler, error) {
+func (p *gateway) scheduleStatCollectionWithDefaultDelay(ctx context.Context, scheduler *sfxclient.Scheduler, loadedConfig *config.GatewayConfig, multiplexer signalfx.Sink) (context.Context, context.CancelFunc) {
+	// We still want to schedule stat collection so people can debug the server if they want
+	scheduler.Sink = dpsink.Discard
```
Comment (mdubbyap) [inline main.go]: this is too much of a copy pasta of the other one. you don't need to set the sink to discard, it should always be the mux, and you'll never be skipping internal metrics stat keeping.

File: main.go [hunk 5]
```diff
@@ -832,7 +866,11 @@ func (p *gateway) start(ctx context.Context) error {
 	// setup scheduler
 	scheduler := p.setupScheduler(p.config)
 
-	if err := p.setupDebugServer(p.config, p.logger, scheduler); err != nil {
+	// set up defaultDelayScheduler
+	defaultDelayScheduler := p.setupScheduler(p.config)
```
Comment (mdubbyap) [inline main.go]: can't use the same setup routine, it includes the go metrics source stuff, we don't want that

File: main_test.go [hunk 6]
```diff
@@ -206,7 +206,8 @@ const emptyConfig = `
     ],
     "ForwardTo":[
     ],
-    "StatsDelay": "5s"
+    "StatsDelay": "5s",
+	"InternalMetricsReportingDelay": "5s"
```
Comment (mdubbyap) [inline main_test.go]: i would like to see a test where only the expected datapoints come out of that reporter. and they  are of the appropriate prefix, etc.

Comment (krunaljain) [reply main_test.go]: working on it

File: main.go (hunk 3 above)
Comment (krunaljain) [reply main.go]: the log context uses the listener type as an argument. Can I pass some default type to reuse the logger across this loop?

File: main.go [hunk 7]
```diff
@@ -582,9 +615,21 @@ func (p *gateway) scheduleStatCollection(ctx context.Context, scheduler *sfxclie
...
 
-func (p *gateway) setupForwardersAndListeners(ctx context.Context, loader *config.Loader, loadedConfig *config.GatewayConfig, logger log.Logger, scheduler *sfxclient.Scheduler) (signalfx.Sink, map[string]http.Handler, error) {
+func (p *gateway) scheduleStatCollectionWithDefaultDelay(ctx context.Context, scheduler *sfxclient.Scheduler, loadedConfig *config.GatewayConfig, multiplexer signalfx.Sink) (context.Context, context.CancelFunc) {
+	scheduler.ReportingDelayNs = (time.Second * 10).Nanoseconds()
+	finishedContext, cancelFunc := context.WithCancel(ctx)
+	if loadedConfig.InternalMetricsReportingDelayDuration != nil && *loadedConfig.InternalMetricsReportingDelayDuration != 0 {
+		scheduler.Sink = multiplexer
+		scheduler.ReportingDelayNs = loadedConfig.InternalMetricsReportingDelayDuration.Nanoseconds()
+	} else {
+		p.logger.Log("skipping internal metrics stat keeping")
```
Comment (mdubbyap) [inline main.go]: again, this isn't a thing.  this is not a skippable thing. If it's not configured it's 10 seconds.

File: main.go [hunk 8]
```diff
@@ -832,7 +875,11 @@ func (p *gateway) start(ctx context.Context) error {
 	// setup scheduler
 	scheduler := p.setupScheduler(p.config)
 
-	if err := p.setupDebugServer(p.config, p.logger, scheduler); err != nil {
+	// set up defaultDelayScheduler
+	defaultDelayScheduler := p.setupDefaultDelayScheduler(p.config)
+	defaultDelayScheduler.Prefix = gatewayMetricsPrefix
```
Comment (mdubbyap) [inline main.go]: this can be set in the same function above as well. we don't need 3 places to configure a scheduler.

**Comment by krunaljain** [comment]
> Am i missing something or is the test that does end to end testing that we were working on not here?

main_test.go line no. 495

File: main_test.go [hunk 9]
```diff
@@ -549,6 +594,80 @@ func TestProxy1(t *testing.T) {
...
+			p.logger = logger
+			loadedConfig, _ := loadConfig(filename, logger)
+			p.configure(loadedConfig)
+			mainDoneChan = make(chan error)
+			go func() {
+				mainDoneChan <- p.start(ctx)
+				close(mainDoneChan)
+			}()
+			<-p.setupDoneSignal
+		}
```
Comment (krunaljain) [inline main_test.go]: Getting 
http: panic serving 127.0.0.1:49484: runtime error: invalid memory address or nil pointer dereference 
after running this test

File: protocol/signalfx/signalfxlistener.go [hunk 10]
```diff
@@ -294,10 +290,7 @@ func SetupChain(ctx context.Context, sink Sink, chainType string, getReader func
 	}
 	st := &sfxclient.WithDimensions{
 		Collector: sfxclient.NewMultiCollector(
-			&metricTracking,
-			&errorTracker,
 			counter,
```
Comment (krunaljain) [inline protocol/signalfx/signalfxlistener.go]: Just making sure if we are okay with deleting these metrics?

Comment (mdubbyap) [reply protocol/signalfx/signalfxlistener.go]: No this is not what we want to do.  We want to exclude the counter from this multi collector, not delete the contents and keep just the counter.  Explore that code and figure out what is actually going on, then come to me and we can discuss it.

**Comment by krunaljain** [comment]
@mdubbyap  can you take a look when you have time? thanks

File: main_test.go [hunk 11]
```diff
@@ -963,3 +966,168 @@ func Test_handleClusterNameResponse(t *testing.T) {
...
+		sfxGateway.signalChan <- syscall.SIGTERM
+		So(sl.Close(), ShouldBeNil)
+		if cancelfunc != nil {
+			cancelfunc()
+		}
+		dps := <-sendTo.PointsChan
+		So(len(dps), ShouldNotEqual, 0)
+		for _, dp := range dps {
+			_, ok := prefixedMetrics[dp.Metric]
```
Comment (mdubbyap) [inline main_test.go]: not something you need to fix, but just education, you're using a map[string]bool, but then ignoring the bool and just checking the existence bool... kinda weird?  it is testing what you think it is, just weird.

**Comment by mdubbyap** [comment]
I approve the code, but please squash the commits into one commit with a concise message about what was done.

File: main_test.go (hunk 11 above)
Comment (krunaljain) [reply main_test.go]: golang doesn't have built in set type and I wanted O(1) lookup. so ended up using a map with bool

**Pull request merged**